File: server/scripts/import-gc.js
/**
 * 将GC数据从JSON导入MySQL
 */
const mysql = require('mysql2/promise');
const fs = require('fs');
const { v4: uuidv4 } = require('uuid');
const dbConfig = require('../config/database');

async function main() {
  let conn;
  try {
    // 读取GC数据
    const gcData = JSON.parse(fs.readFileSync('./stage5-gc.json', 'utf-8'));
    console.log(`🚴 GC数据导入工具`);
    console.log(`📊 读取到 ${gcData.length} 条GC记录\n`);
    
    conn = await mysql.createConnection({
      ...dbConfig.development,
      database: dbConfig.development.database
    });
    
    // 获取race_id和stage_id
    const [races] = await conn.query('SELECT * FROM races WHERE race_code = ?', ['giro-ditalia-2026']);
    if (races.length === 0) {
      console.log('❌ 赛事不存在');
      return;
    }
    const raceId = races[0].id;
    
    const [stages] = await conn.query(
      'SELECT * FROM stages WHERE race_id = ? AND stage_number = ?',
      [raceId, 5]
    );
    if (stages.length === 0) {
      console.log('❌ 赛段不存在');
      return;
    }
    const stageId = stages[0].id;
    
    console.log(`✅ 赛事ID: ${raceId}`);
    console.log(`✅ 赛段ID: ${stageId}\n`);
    
    // 导入GC数据
    console.log('📊 导入GC总成绩榜...\n');
    
    let imported = 0;
    let skipped = 0;
    
    for (const gc of gcData) {
      try {
        // 查找或创建车手
        const [riders] = await conn.query('SELECT * FROM riders WHERE rider_name = ?', [gc.rider_name]);
        let riderId;
        if (riders.length > 0) {
          riderId = riders[0].id;
        } else {
          riderId = uuidv4();
          await conn.query(
            'INSERT INTO riders (id, rider_name, nationality) VALUES (?, ?, ?)',
            [riderId, gc.rider_name, 'UNK']
          );
        }
        
        // 查找或创建车队
        const [teams] = await conn.query('SELECT * FROM teams WHERE team_name = ?', [gc.team_name]);
        let teamId;
        if (teams.length > 0) {
          teamId = teams[0].id;
        } else {
          teamId = uuidv4();
          await conn.query('INSERT INTO teams (id, team_name) VALUES (?, ?)', [teamId, gc.team_name]);
        }
        
        // 插入GC数据
        await conn.query(`
          INSERT INTO general_classification (id, stage_id, \`rank\`, rider_id, team_id, total_time, time_gap)
          VALUES (?, ?, ?, ?, ?, ?, ?)
          ON DUPLICATE KEY UPDATE
            rider_id = VALUES(rider_id),
            team_id = VALUES(team_id),
            total_time = VALUES(total_time),
            time_gap = VALUES(time_gap)
        `, [uuidv4(), stageId, gc.rank, riderId, teamId, gc.total_time || null, gc.time_gap]);
        
        imported++;
        if (imported <= 10 || imported % 20 === 0) {
          console.log(`  ✅ ${gc.rank}. ${gc.rider_name} (${gc.team_name}) - ${gc.time_gap}`);
        }
      } catch (err) {
        skipped++;
        if (skipped <= 5) {
          console.error(`  ❌ 失败 [${gc.rank}] ${gc.rider_name}:`, err.message);
        }
      }
    }
    
    console.log(`\n📊 GC导入完成: ${imported} 成功, ${skipped} 失败\n`);
    
    // 验证
    const [count] = await conn.query('SELECT COUNT(*) as count FROM general_classification WHERE stage_id = ?', [stageId]);
    console.log(`✅ GC表中该赛段共有 ${count[0].count} 条记录\n`);
    
    // 查询前10
    const [top10] = await conn.query(`
      SELECT gc.rank, r.rider_name, t.team_name, gc.time_gap, gc.total_time
      FROM general_classification gc
      JOIN riders r ON gc.rider_id = r.id
      JOIN teams t ON gc.team_id = t.id
      WHERE gc.stage_id = ?
      ORDER BY gc.rank
      LIMIT 10
    `, [stageId]);
    
    console.log('🏆 GC总成绩榜前10：');
    console.log('排名 | 车手 | 车队 | 时间差 | 总时间');
    console.log('-'.repeat(100));
    top10.forEach(r => {
      console.log(`${String(r.rank).padEnd(6)} | ${r.rider_name.padEnd(25)} | ${r.team_name.padEnd(30)} | ${(r.time_gap || '').padEnd(10)} | ${r.total_time || ''}`);
    });
    
    console.log('\n🎉 GC数据导入完成！');
    
  } catch (err) {
    console.error('❌ 失败:', err);
  } finally {
    if (conn) await conn.end();
  }
}

main();

File: server/scripts/import_gc.py
#!/usr/bin/env python3
"""
导入GC总成绩数据到 general_classification 表
"""

import json
import pymysql
import sys

# 数据库连接配置 - 本地开发环境
DB_CONFIG = {
    'host': '127.0.0.1',
    'port': 13306,
    'user': 'root',
    'password': 'mysql123456',
    'database': 'jersey_db',
    'charset': 'utf8mb4'
}

def get_connection():
    return pymysql.connect(**DB_CONFIG)

def import_gc_data(json_file, race_id, stage_num):
    """导入GC数据到数据库"""
    
    # 读取JSON文件
    with open(json_file, 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    gc_data = data.get('gc', [])
    print(f"准备导入GC数据: {len(gc_data)} 条")
    
    if not gc_data:
        print("警告: 没有GC数据")
        return
    
    conn = get_connection()
    cursor = conn.cursor()
    
    try:
        # 查找stage_id
        cursor.execute(
            "SELECT id FROM stages WHERE race_id = %s AND stage_number = %s",
            (race_id, stage_num)
        )
        stage_result = cursor.fetchone()
        
        if not stage_result:
            print(f"错误: 找不到 Stage {stage_num}")
            return
        
        stage_id = stage_result[0]
        print(f"Stage ID: {stage_id}")
        
        # 删除该赛段已有的GC数据
        cursor.execute("DELETE FROM general_classification WHERE stage_id = %s", (stage_id,))
        deleted = cursor.rowcount
        if deleted > 0:
            print(f"删除旧GC数据: {deleted} 条")
        
        # 导入新的GC数据
        inserted = 0
        for gc in gc_data:
            rank = int(gc.get('rank', 0))
            rider_name = gc.get('rider', '')
            team_name = gc.get('team', '')
            time_gap = gc.get('time', '')  # GC时间差
            
            if not rider_name or not rank:
                continue
            
            # 查找rider_id
            cursor.execute("SELECT id FROM riders WHERE rider_name = %s", (rider_name,))
            rider_result = cursor.fetchone()
            
            if not rider_result:
                print(f"  警告: 找不到车手 {rider_name}，跳过")
                continue
            
            rider_id = rider_result[0]
            
            # 查找team_id
            cursor.execute("SELECT id FROM teams WHERE team_name = %s", (team_name,))
            team_result = cursor.fetchone()
            
            if not team_result:
                print(f"  警告: 找不到车队 {team_name}，使用空ID")
                team_id = ''
            else:
                team_id = team_result[0]
            
            # 插入GC数据
            cursor.execute("""
                INSERT INTO general_classification (id, stage_id, `rank`, rider_id, team_id, nationality, time_gap)
                VALUES (UUID(), %s, %s, %s, %s, %s, %s)
            """, (
                stage_id,
                rank,
                rider_id,
                team_id,
                gc.get('nationality', ''),
                time_gap
            ))
            
            inserted += 1
            
            if inserted % 50 == 0:
                print(f"  已处理 {inserted}/{len(gc_data)}...")
        
        conn.commit()
        print(f"\n✓ GC数据导入完成: {inserted} 条")
        
    except Exception as e:
        conn.rollback()
        print(f"错误: {e}")
        import traceback
        traceback.print_exc()
    finally:
        cursor.close()
        conn.close()

if __name__ == '__main__':
    json_file = 'stage-16-complete.json'
    race_id = 'e597183f-8ea4-4fb0-a469-661c57c5a958'  # 2026男子环意
    stage_num = 16
    
    import_gc_data(json_file, race_id, stage_num)
